### PR TITLE
feat: control indexing in temporal extraction

### PR DESCRIPTION
**Pull request by rok**
BREAKING CHANGE: This adds an option to control indexing of components as discussed in #477.

### CODE REVIEW

File: extensions/functions_datetime.yaml
Comment (rok) [reply extensions/functions_datetime.yaml]: Done.

File: extensions/functions_datetime.yaml
Comment (rok) [reply extensions/functions_datetime.yaml]: Changed.

**Comment by ianmcook** [comment]
@rok I reworked the documentation to be clearer. Could you please review this, and if you agree that it's clearer this way, please replace it with this? Thanks.
```
      Extract portion of a date/time value.
      * YEAR Return the year.
      * ISO_YEAR Return the ISO 8601 week-numbering year. First week of an ISO year has the majority (4 or more) of
        its days in January.
      * US_YEAR Return the US epidemiological year. First week of US epidemiological year has the majority (4 or more)
        of its days in January. Last week of US epidemiological year has the year's last Wednesday in it. US 
        epidemiological week starts on Sunday.
      * QUARTER Return the number of the quarter within the year. January 1 through March 31 map to the first quarter,
        April 1 through June 30 map to the second quarter, etc.
      * MONTH Return the number of the month within the year.
      * DAY Return the number of the day within the month.
      * DAY_OF_YEAR Return the number of the day within the year. January 1 maps to the first day, February 1 maps to
        the thirty-second day, etc.
      * MONDAY_DAY_OF_WEEK Return the number of the day within the week, from Monday (first day) to Sunday (seventh
        day).
      * SUNDAY_DAY_OF_WEEK Return the number of the day within the week, from Sunday (first day) to Saturday (seventh
        day).
      * MONDAY_WEEK Return the number of the week within the year. First week starts on first Monday of January.
      * SUNDAY_WEEK Return the number of the week within the year. First week starts on first Sunday of January.
      * ISO_WEEK Return the number of the ISO week within the ISO year. First ISO week has the majority (4 or more)
        of its days in January. ISO week starts on Monday.
      * US_WEEK Return the number of the US week within the ISO year. First US week has the majority (4 or more) of
        its days in January. US week starts on Sunday.
      * HOUR Return the hour (0-23).
      * MINUTE Return the minute (0-59).
      * SECOND Return the second (0-59).
      * MILLISECOND Return number of milliseconds since the last full second.
      * MICROSECOND Return number of microseconds since the last full millisecond.
      * SUBSECOND Return number of microseconds since the last full second of the given timestamp.
      * UNIX_TIME Return number of seconds that have elapsed since 1970-01-01 00:00:00 UTC, ignoring leap seconds.
      * TIMEZONE_OFFSET Return number of seconds of timezone offset to UTC.

      The range of values returned for QUARTER, MONTH, DAY, DAY_OF_YEAR, MONDAY_DAY_OF_WEEK, SUNDAY_DAY_OF_WEEK,
      MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, and US_WEEK depends on whether counting starts at 1 or 0. This is governed
      by the indexing option.

      When indexing is ONE:
       * QUARTER returns values in range 1-4
       * MONTH returns values in range 1-12
       * DAY returns values in range 1-31
       * DAY_OF_YEAR returns values in range 1-366
       * MONDAY_DAY_OF_WEEK and SUNDAY_DAY_OF_WEEK return values in range 1-7
       * MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, and US_WEEK return values in range 1-53

      When indexing is ZERO:
       * QUARTER returns values in range 0-3
       * MONTH returns values in range 0-11
       * DAY returns values in range 0-30
       * DAY_OF_YEAR returns values in range 0-365
       * MONDAY_DAY_OF_WEEK and SUNDAY_DAY_OF_WEEK return values in range 0-6
       * MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, and US_WEEK return values in range 0-52

       Timezone strings must be as defined by IANA timezone database (https://www.iana.org/time-zones).
       Examples: "Pacific/Marquesas", "Etc/GMT+1".
       If timezone is invalid an error is thrown.
```

**Comment by ianmcook** [comment]
@rok one thing I was unclear about is whether:

`US_WEEK Return the number of the US week within the ISO year.`
or
`US_WEEK Return the number of the US week within the US year. `

Please update the docs to reflect whichever is correct.

**Review by ianmcook — DISMISSED** [review]

File: extensions/functions_datetime.yaml
Comment (rok) [reply extensions/functions_datetime.yaml]: But the fact we're adding an option makes this a breaking change, correct?

Comment (ianmcook) [reply extensions/functions_datetime.yaml]: Yes

**Review by westonpace — APPROVED** [review]
+1

Since it is a breaking change we will need further approval.  @jacques-n / @cpcloud ?

File: extensions/functions_datetime.yaml
Comment (ianmcook) [reply extensions/functions_datetime.yaml:71]: Thanks for pointing that out @jorisvandenbossche. I agree with @rok that it would be best to create separate `impls` entries for the cases in which `indexing` has no effect.

Here's my first stab at that. Four implementations that do not require `indexing` followed by three implementations that do require `indexing`. Does this look OK?

```yaml
    impls:
      - args:
          - name: component
            options: [ YEAR, ISO_YEAR, US_YEAR, HOUR, MINUTE, SECOND,
                       MILLISECOND, MICROSECOND, SUBSECOND, UNIX_TIME, TIMEZONE_OFFSET ]
            description: The part of the value to extract.
          - name: x
            value: timestamp_tz
          - name: timezone
            description: Timezone string from IANA tzdb.
            value: string
        return: i64
      - args:
          - name: component
            options: [ YEAR, ISO_YEAR, US_YEAR, HOUR, MINUTE, SECOND,
                       MILLISECOND, MICROSECOND, SUBSECOND, UNIX_TIME ]
            description: The part of the value to extract.
          - name: x
            value: timestamp
        return: i64
      - args:
          - name: component
            options: [ YEAR, ISO_YEAR, US_YEAR, UNIX_TIME ]
            description: The part of the value to extract.
          - name: x
            value: date
        return: i64
      - args:
         - name: component
           options: [ HOUR, MINUTE, SECOND, MILLISECOND, MICROSECOND, SUBSECOND ]
           description: The part of the value to extract.
         - name: x
           value: time
        return: i64
      - args:
          - name: component
            options: [ QUARTER, MONTH, DAY, DAY_OF_YEAR, MONDAY_DAY_OF_WEEK,
                       SUNDAY_DAY_OF_WEEK, MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, US_WEEK ]
            description: The part of the value to extract.
          - name: indexing
            options: [ ONE, ZERO ]
            description: Start counting from 1 or 0.
          - name: x
            value: timestamp_tz
          - name: timezone
            description: Timezone string from IANA tzdb.
            value: string
        return: i64
      - args:
          - name: component
            options: [ QUARTER, MONTH, DAY, DAY_OF_YEAR, MONDAY_DAY_OF_WEEK,
                       SUNDAY_DAY_OF_WEEK, MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, US_WEEK ]
            description: The part of the value to extract.
          - name: indexing
            options: [ ONE, ZERO ]
            description: Start counting from 1 or 0.
          - name: x
            value: timestamp
        return: i64
      - args:
          - name: component
            options: [ QUARTER, MONTH, DAY, DAY_OF_YEAR, MONDAY_DAY_OF_WEEK,
                       SUNDAY_DAY_OF_WEEK, MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, US_WEEK ]
            description: The part of the value to extract.
          - name: indexing
            options: [ ONE, ZERO ]
            description: Start counting from 1 or 0.
          - name: x
            value: date
        return: i64
```

Comment (rok) [reply extensions/functions_datetime.yaml]: Thanks @ianmcook. I pasted your change into the last commit.

File: extensions/functions_datetime.yaml
Comment (ianmcook) [inline extensions/functions_datetime.yaml:58]: Suggested docs improvement to accompany the implementation change:
```suggestion

      The indexing option must be specified when the component is QUARTER, MONTH, DAY, DAY_OF_YEAR,
      MONDAY_DAY_OF_WEEK, SUNDAY_DAY_OF_WEEK, MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, or US_WEEK. The 
      indexing option cannot be specified when the component is YEAR, ISO_YEAR, US_YEAR, HOUR, MINUTE, SECOND,
      MILLISECOND, MICROSECOND, SUBSECOND, UNIX_TIME, or TIMEZONE_OFFSET.

```

**Review by ianmcook — DISMISSED** [review]
LGTM. The YAML has some formatting issues, but I believe that will get automatically fixed after merge by the CI.

**Review by ianmcook — APPROVED** [review]

**Review by westonpace — DISMISSED** [review]
@ianmcook what formatting issues?  We don't have any auto-formatting in the CI.

My approval remains.  But we will still need further approval since it is a breaking change.

**Comment by ianmcook** [comment]
> @ianmcook what formatting issues?

Just very minor stuff like inconsistency in the number of leading spaces used for indenting the description

File: extensions/functions_datetime.yaml
Comment (ianmcook) [inline extensions/functions_datetime.yaml:66]: There's one space too much indentation on this paragraph
```suggestion
      Timezone strings must be as defined by IANA timezone database (https://www.iana.org/time-zones).
      Examples: "Pacific/Marquesas", "Etc/GMT+1".
      If timezone is invalid an error is thrown.
```

File: extensions/functions_datetime.yaml
Comment (ianmcook) [inline extensions/functions_datetime.yaml:57]: There is an extra leading space before the bullet characters here compared to in the bullets above
```suggestion
      When indexing is ONE:
      * QUARTER returns values in range 1-4
      * MONTH returns values in range 1-12
      * DAY returns values in range 1-31
      * DAY_OF_YEAR returns values in range 1-366
      * MONDAY_DAY_OF_WEEK and SUNDAY_DAY_OF_WEEK return values in range 1-7
      * MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, and US_WEEK return values in range 1-53

      When indexing is ZERO:
      * QUARTER returns values in range 0-3
      * MONTH returns values in range 0-11
      * DAY returns values in range 0-30
      * DAY_OF_YEAR returns values in range 0-365
      * MONDAY_DAY_OF_WEEK and SUNDAY_DAY_OF_WEEK return values in range 0-6
      * MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, and US_WEEK return values in range 0-52
```

**Comment by westonpace** [comment]
@rok did you want to merge Ian's formatting changes?

**Comment by rok** [comment]
Done. Sorry for the wait, I had a busy week.

**Review by ianmcook — APPROVED** [review]

**Review by westonpace — APPROVED** [review]
No problem.  I wasn't rushed, just meant that as a friendly ping.  Thanks for taking care of this.

**Comment by westonpace** [comment]
@jacques-n / @cpcloud second ping for a second +1 so we can merge this.

**Comment by cpcloud** [comment]
+1

**Pull request merged**
